### PR TITLE
[WIP] Refactor and implement ChangeDataHolderEvent.ValueChange

### PR DESCRIPTION
**Pull request by Aaron1011**
[SpongeAPI](https://github.com/SpongePowered/SpongeAPI/pull/1695) | **SpongeCommon** | [SpongeForge](https://github.com/SpongePowered/SpongeForge/pull/1866)

See the SpongeAPI PR for more details.

This is the basic pattern for firing a `ChangeDataHolderEvent.ValueChange` event for a given key:

1. Redirect all locations in Vanilla/Forge where the value is set. Ideally, there will be a public setter method for you to inject into.
2. Instead of setting the underlying field (or `DataParameter, etc`), call `offerWithEvent` on the appropriate `DataHolder`, putting any relevant objects into the `Cause`
3. In the `ValueProcessor` for the key in question, **bypass** the setter you redirected before, if applicable. This ensures that setting the value does not end up recursing infinitely between your redirected setter and the value processor.

Basically, the idea is to ensure that all changes done by mods/vanilla/forge end up firing an event, while API changes (e.g. anything going through the value processor) does **not**. The decision as to whether or not to fire an event for a plugin change is handled on the API side, and is not the implementation's concern.

### CODE REVIEW

File: src/main/java/org/spongepowered/common/data/AbstractArchetype.java [hunk 1]
```diff
@@ -122,6 +124,11 @@ public boolean supports(Class<? extends DataManipulator<?, ?>> holderClass) {
                 .orElseGet(DataTransactionResult::failNoData);
     }
 
+    @Override
+    public <E> Optional<ChangeDataHolderEvent.ValueChange> offerWithEvent(Key<? extends BaseValue<E>> key, E value, Cause cause) {
+        throw new UnsupportedOperationException();
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/AbstractArchetype.java:129]: Message maybe?

File: src/main/java/org/spongepowered/common/data/AbstractArchetype.java [hunk 2]
```diff
@@ -192,12 +199,24 @@ public DataTransactionResult copyFrom(DataHolder that, MergeFunction function) {
                 .flatMap(processor -> processor.readValue(this.data));
     }
 
+    @SuppressWarnings("unchecked")
+    @Override
+    public <R> Optional<R> getDefault(Key<? extends BaseValue<R>> key) {
+        return DataUtil.getNbtProcessor(this.getDataType(), key)
+                .flatMap(processor -> processor.readFrom(this.data).map(v -> v.getDefault()));
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/AbstractArchetype.java:206]: multi-line the nested mapping. In the event there's an exception, we want to be able to pinpoint which lambda is having an issue.

File: src/main/java/org/spongepowered/common/data/DataProcessor.java [hunk 3]
```diff
@@ -97,6 +98,8 @@
      */
     DataTransactionResult set(DataHolder dataHolder, M manipulator, MergeFunction function);
 
+    //void setWithEvent(DataHolder container, M manipulator, Cause cause);
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/DataProcessor.java:101]: Intended to be used later maybe? comments welcome.

File: src/main/java/org/spongepowered/common/data/processor/common/AbstractBlockOnlyDataProcessor.java [hunk 4]
```diff
@@ -25,13 +25,16 @@
...
+import org.spongepowered.api.data.DataHolder;
 import org.spongepowered.api.data.DataTransactionResult;
 import org.spongepowered.api.data.key.Key;
 import org.spongepowered.api.data.manipulator.DataManipulator;
 import org.spongepowered.api.data.manipulator.ImmutableDataManipulator;
 import org.spongepowered.api.data.value.BaseValue;
 import org.spongepowered.api.data.value.ValueContainer;
 import org.spongepowered.api.data.value.immutable.ImmutableValue;
+import org.spongepowered.api.event.cause.Cause;
+import org.spongepowered.api.event.data.ChangeDataHolderEvent;
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/common/AbstractBlockOnlyDataProcessor.java:37]: imports?

File: src/main/java/org/spongepowered/common/mixin/core/util/MixinFoodStats.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+    @Shadow private int foodLevel;
+    private Player player; // Null when on client
+
+    @Override
+    public void setPlayer(EntityPlayer player) {
+        this.player = (Player) player;
+    }
+
+    @Inject(method = "addStats(Lnet/minecraft/item/ItemFood;Lnet/minecraft/item/ItemStack;)V", at = @At(value = "HEAD"))
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/core/util/MixinFoodStats.java:55]: This should be a redirect if you're modifying the cause stack, this way you can wrap it in a try with autocloseable frames.

File: testplugins/src/main/java/org/spongepowered/test/FoodChangeTest.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+import org.spongepowered.api.event.filter.cause.First;
+import org.spongepowered.api.event.filter.data.GetKey;
+import org.spongepowered.api.plugin.Plugin;
+import org.spongepowered.api.text.Text;
+import org.spongepowered.api.text.channel.MessageChannel;
+
+import java.util.Optional;
+
+@Plugin(id = "food-change-test", authors = "Aaron1011")
+public class FoodChangeTest {
```
Comment (gabizou) [inline testplugins/src/main/java/org/spongepowered/test/FoodChangeTest.java:23]: If this is going to constantly print to the player, should make it a toggle to join in (reduces the amount of commenting when debugging implementation).

File: testplugins/src/main/java/org/spongepowered/test/InjectionTest.java [hunk 7]
```diff
@@ -25,6 +25,8 @@
 package org.spongepowered.test;
 
 import org.slf4j.Logger;
+import org.spongepowered.api.data.key.Keys;
```
Comment (gabizou) [inline testplugins/src/main/java/org/spongepowered/test/InjectionTest.java:28]: imports.

**Review by gabizou — CHANGES_REQUESTED** [review]
Small changes, otherwise looks good. Aside from API discussions we've had about removing GetKey annotation.

**Comment by parlough** [comment]
Going to close this for now due to https://github.com/SpongePowered/SpongeAPI/pull/1728 and https://github.com/SpongePowered/SpongeCommon/pull/1670 which provide a method of implementing and listening to the event. 

If you'd like to make other changes, open a new PR with those in relation to what is now in the API and implementation.
